src/Stmt.h

```diff
@@ -481,23 +481,18 @@ class StmtList : public Stmt
 	bool NoFlowAfter(bool ignore_break) const override;
 
 	// Idioms commonly used in reduction.
-	StmtList(StmtPtr s1, Stmt* s2);
 	StmtList(StmtPtr s1, StmtPtr s2);
 	StmtList(StmtPtr s1, StmtPtr s2, StmtPtr s3);
 
 protected:
 	bool IsPure() const override;
 
-	StmtPList* stmts;
+	std::vector<StmtPtr> stmts;
 
 	// Optimization-related:
-	bool ReduceStmt(int& s_i, StmtPList* f_stmts, Reducer* c);
+	bool ReduceStmt(int& s_i, std::vector<StmtPtr>& f_stmts, Reducer* c);
 
-	void ResetStmts(StmtPList* new_stmts)
-		{
-		delete stmts;
-		stmts = new_stmts;
-		}
+	void ResetStmts(std::vector<StmtPtr>& new_stmts) { stmts = std::move(new_stmts); }
```

Comment:
Nothing to address, just staring ad references and std::move(). This is a bit of a funky API because the caller's vector is valid but unspecified (concretely empty) after a call which could be surprising.

It's not part of the public API (protected) and only used in script_opt so no reason to address, but wondering if a value arg may be good enough or if this is so performance critical? Anyhow, nothing to do.

Reply:
I guess I don't have a good mental model of just what `std::move` does. I would think that above it does a copy (and then clears `new_stmts`, so if I passed `new_stmts` as a non-reference there would be two copies, one onto the call stack and one from it. Is that not right?

Regarding performance critical, that method does indeed get banged on a lot, every time a sequence of statements changes while transforming the AST to reduced form, and because some of the transformations compound, this is effectively done over and over until there are no more changes.

Reply:
`std::move` is just a cast from e.g., `T` to a `T&&`. In this case a `T&&` on the RHS of the assignment allows `stmts` to directly take over the ownership of the backing storage behind `new_stmts` so that the assignment is just pointer assignments inside `vector`. After that `new_stmts` is a in a "valid but undefined state" so you cannot really do much with it anymore but e.g., assign to it again.

I believe moving from the arg is a good idea here, but since you very likely do not need the moved from arg in the calling scope, so I would instead accept a `std::vector<StmtPtr>` instead of a reference, and  `std::move` the arg at the call sites. That `std::move` will cause creation of a arg value from an rvalue which does the same efficient internal pointer assignments without deep copy like at the `std::move` here.

Reply:
> If a different name would help with the API funkiness, I'm fine with that. Haven't been able to come up with one myself. (Or I could just add a heads-up comment, though it's sort of directly apparent if you're looking at the declaration given the inlined definition.)

I think it's fine due to the special use, detail namespace and protected. Maybe trying Benjamin's suggestion, but the comment mainly stemmed from me derailing into "what if this was a public API / commonly used API".

Reply:
Thanks for the sketch - just pushed the better approach.